pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1393,6 +1452,415 @@ public PartitionAssignment getStreamPartitionAssignment(String realtimeTableName
+    }
+
+    for (Map.Entry<String, String> entry : segmentMapper.getAll().entrySet()) {
+      String oldSeg = entry.getKey();
+      String newSeg = entry.getValue();
+      List<String> newInstances = assignments.get(newSeg);
+      Map<String, String> instanceStateMap = idealState.getInstanceStateMap(newSeg);
+      if (instanceStateMap != null) {
+        instanceStateMap.clear();
+      }
+      for (String instance : newInstances) {
+        idealState.setPartitionState(newSeg, instance,
+            PinotHelixSegmentOnlineOfflineStateModelGenerator.CONSUMING_STATE);
+      }
+      if (!segmentMapper.isNewSegmentKey(oldSeg)) {
+        Set<String> oldInstances = idealState.getInstanceSet(oldSeg);
+        Preconditions.checkArgument(CollectionUtils.isNotEmpty(oldInstances));
+        for (String instance : oldInstances) {
+          idealState.setPartitionState(oldSeg, instance,
+              PinotHelixSegmentOnlineOfflineStateModelGenerator.ONLINE_STATE);
+        }
+      }
+    }
+    return idealState;
+  }
+
+  @VisibleForTesting
+  protected long getCurrentTimeMs() {
+    return System.currentTimeMillis();
+  }
+
+  protected IdealState updateIdealStateOnSegmentCompletion(@Nonnull IdealState idealState, @Nonnull String currentSegmentId,
+      @Nonnull String newSegmentId, @Nonnull  PartitionAssignment partitionAssignment) {
+
+    Map<String, List<String>> instanceAssignments = null;
+
+    RealtimeSegmentAssignmentStrategy strategy = new ConsumingSegmentAssignmentStrategy();
+    try {
+      instanceAssignments = strategy.assign(Lists.newArrayList(newSegmentId), partitionAssignment);
+    } catch (InvalidConfigException e) {
+      _controllerMetrics.addMeteredTableValue(idealState.getResourceName(),
+          ControllerMeter.CONTROLLER_REALTIME_TABLE_SEGMENT_ASSIGNMENT_ERROR, 1L);
+      throw new IllegalStateException("Caught exception when updating ideal state on segment completion", e);
+    }
+
+    List<String> newSegmentInstances = instanceAssignments.get(newSegmentId);
+    Set<String> currentSegmentInstances = idealState.getInstanceSet(currentSegmentId);
+    for (String instance : currentSegmentInstances) {
+      idealState.setPartitionState(currentSegmentId, instance,
+          PinotHelixSegmentOnlineOfflineStateModelGenerator.ONLINE_STATE);
+    }
+
+    // We may have (for whatever reason) a different instance list in the idealstate for the new segment.
+    // If so, clear it, and then set the instance state for the set of instances that we know should be there.
+    Map<String, String> stateMap = idealState.getInstanceStateMap(newSegmentId);
+    if (stateMap != null) {
+      stateMap.clear();
+    }
+    for (String instance : newSegmentInstances) {
+      idealState.setPartitionState(newSegmentId, instance,
+          PinotHelixSegmentOnlineOfflineStateModelGenerator.CONSUMING_STATE);
+    }
+
+    return idealState;
+  }
+
+  /**
+   * Class to maintain the mapping from old segments to new segments
```

Comment:
can u add more comments here?the class really keeps a mapping of  segment names that need to move from CONSUMING to ONLINE,   to new segment names (to be created in CONSUMING state). In some cases, we will not have old segment names (e.g. when a new partition is detected, or the old segment is OFFLINE).

Maybe we can keep two different sets instead? One that keeps the old segment that need to move to ONLINE and another that keeps new segments that need to be created in CONSUMING mode. Then the logic is more readable.